nestkernel/source_table.cpp

```diff
@@ -408,3 +448,103 @@ nest::SourceTable::get_next_target_data( const thread tid,
     return true; // found a valid entry
   }
 }
+
+void
+nest::SourceTable::resize_compressible_sources()
+{
+  for ( thread tid = 0; tid < static_cast< thread >( compressible_sources_.size() ); ++tid )
+  {
+    compressible_sources_[ tid ].clear();
+    compressible_sources_[ tid ].resize(
+      kernel().model_manager.get_num_synapse_prototypes(), std::map< index, SpikeData >() );
+  }
+}
+
+void
+nest::SourceTable::collect_compressible_sources( const thread tid )
+{
+  for ( synindex syn_id = 0; syn_id < sources_[ tid ].size(); ++syn_id )
+  {
+    index lcid = 0;
+    while ( lcid < sources_[ tid ][ syn_id ].size() )
+    {
+      const std::pair< index, SpikeData > source_node_id_to_spike_data =
+        std::make_pair( sources_[ tid ][ syn_id ][ lcid ].get_node_id(), SpikeData( tid, syn_id, lcid, 0 ) );
+      compressible_sources_[ tid ][ syn_id ].insert( source_node_id_to_spike_data );
+
+      // find next source with different node_id (assumes sorted sources)
+      const index old_source_node_id = sources_[ tid ][ syn_id ][ lcid ].get_node_id();
+      ++lcid;
+      while ( ( lcid < sources_[ tid ][ syn_id ].size() )
+        and ( sources_[ tid ][ syn_id ][ lcid ].get_node_id() == old_source_node_id ) )
+      {
+        ++lcid;
+      }
+    }
+  }
+}
+
+void
+nest::SourceTable::fill_compressed_spike_data(
+  std::vector< std::vector< std::vector< SpikeData > > >& compressed_spike_data )
+{
+  compressed_spike_data.clear();
+  compressed_spike_data.resize( kernel().model_manager.get_num_synapse_prototypes() );
+
+  for ( thread tid = 0; tid < static_cast< thread >( compressible_sources_.size() ); ++tid )
```

Comment:
Replace by `#pragma omp parallel`?

Reply:
can't, we're in an omp single block https://github.com/jakobj/nest-simulator/blob/enh/com-com/nestkernel/connection_manager.cpp#L1494

Reply:
OK!